lib/web/sessions.go

```diff
@@ -503,6 +503,47 @@ func (c *SessionContext) GetUserAccessChecker() (services.AccessChecker, error)
 	return accessChecker, trace.Wrap(err)
 }
 
+// GetUserAccessChecker returns a [services.AccessChecker] derived from the SSH certificate
+// associated with this session for the provided cluster.
+func (c *SessionContext) GetUserAccessChecker(ctx context.Context, site reversetunnelclient.RemoteSite) (services.AccessChecker, error) {
+	if c.cfg.RootClusterName == site.GetName() {
+		checker, err := c.GetAccessChecker()
+		return checker, trace.Wrap(err)
+	}
+
+	clt, err := site.CachingAccessPoint()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get the UserCA of the root cluster from the leaf cluster. This ensures
+	// the role mappings are populated according to the trusted cluster in the
+	// leaf cluster.
+	ca, err := clt.GetCertAuthority(ctx, types.CertAuthID{
+		Type:       types.UserCA,
+		DomainName: c.cfg.RootClusterName,
+	}, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cert, err := c.GetSSHCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	info, err := services.AccessInfoFromRemoteCertificate(cert, ca.CombinedMapping())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The leaf cluster client is provided to the access checker so that roles
+	// are retrieved from the leaf and not the root. This allows the access checker
+	// to know about any logins defined in the leaf cluster.
+	accessChecker, err := services.NewAccessChecker(info, site.GetName(), clt)
```

Comment:
I know we already use a similar approach in other places and it works

but my gut reaction is to question why we are doing access checks on the proxy on behalf of leaf clusters. In general I don't think we can really presume that the root proxy knows how a leaf would evaluate access. E.g. the root proxy could be on v14, the leaf proxy on v13, and the leaf node on v12 - technically that should be fine but if we start making changes to RBAC I could see the idea of which logins should be allowed on that leaf node drifting. It's not hard to imagine other possible RBAC differences if certain features are enabled in one cluster and not the other, like when we start making changes with TAG

It's probably fine for just listing allowed logins in the UI, but with a generic name like `GetUserAccessChecker` I do worry that we will start doing this more often.

I think in an ideal world we would include these logins in the response from the ListResources endpoint on the leaf auth server so all the access checks happen in one place? I understand that would be a pretty monumental change so I think I'm okay with doing it this way for this feature, but just wanted to put these thoughts out there.

Reply:
Right, it would probably have to be a separate RPC, there's also the unified resource list endpoints to consider now

I just mean that in a hypothetical ideal scenario, the proxy could make a single call to get a list of all resources the user can access with the associated logins that should be displayed, instead of the current scenario where we make the proxy:
- first fetch a page of resources
- then map the user's local roles to remote roles
- then fetch all the user's remote roles from the leaf
- then redo all the access checks with the leaf cluster roles
^ and supporting all this going in to the future with all possible RBAC changes, across versions, I am not confident we could make the root proxy always arrive at the same access decisions as the leaf cluster

Reply:
Sometimes you don't need a certificate, the proxy can make all access decisions on its own, like with agentless nodes, and it's not entirely clear how those access decisions are spread across root and leaf proxies and auth servers. We have had bugs in this area.

It's probably fine, if a bit of a hack, for displaying the right logins in the UI and tsh output. My concern with this PR is that providing something as that seems as fundamental and innocent as SessionContext.GetUserAccessChecker may make this pattern increasingly pervasive

The access decisions between root and leaf are already drifting in the case of `login_leafcluster` that marco mentioned.

Reply:
I understood what you meant. Showing the logins now may be moot, I mean that if we continue into a future where GetUserAccessChecker is more prevalent and root proxies are doing the access checks for leaf nodes, especially agentless nodes, lack of a cert with the login may not stop anything, the root proxy might be granting more or less access with or without it

Reply:
If this PR somehow made it more explicit that this style of access check on the proxy should only be used for non-critical UI helpers, I would be much less concerned